flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java

```diff
@@ -55,50 +57,101 @@
  */
 @Internal
 public abstract class KafkaTableSource
-	implements StreamTableSource<Row>, DefinedProctimeAttribute, DefinedRowtimeAttributes {
+	implements StreamTableSource<Row>, DefinedProctimeAttribute, DefinedRowtimeAttributes, DefinedFieldMapping {
+
+	// common table source attributes
+	// TODO make all attributes final once we drop support for format-specific table sources
 
 	/** The schema of the table. */
 	private final TableSchema schema;
 
+	/** Field name of the processing time attribute, null if no processing time field is defined. */
+	private String proctimeAttribute;
+
+	/** Descriptor for a rowtime attribute. */
+	private List<RowtimeAttributeDescriptor> rowtimeAttributeDescriptors;
+
+	/** Mapping for the fields of the table schema to fields of the physical returned type or null. */
+	private Map<String, String> fieldMapping;
```

Comment:
`Optional<Map<...>> fieldMapping`. Same goes for other nullable fields/parameters.

Reply:
I added a `@Nullable` annotation.